python_ta/transforms/type_inference_visitor.py

```diff
@@ -290,6 +298,28 @@ def visit_ifexp(self, node):
         else:
             node.type_constraints = TypeInfo(Any)
 
+    def visit_comprehension(self, node):
+        # TODO: use helper!
+        arg_type = self.type_constraints.lookup_concrete(node.iter.type_constraints.type)
+        iterable_type = self.type_store.lookup_function('__iter__', arg_type)
+        rtype = self.type_constraints.unify_call(iterable_type, arg_type)
+        if isinstance(node.target, astroid.Tuple):
+            for target_node in node.target.elts:
+                target_tvar = node.parent.type_environment.lookup_in_env(target_node.name)
+                self.type_constraints.unify(target_tvar, rtype.__args__[0].__args__[0])
+        else:
+            target_tvar = node.parent.type_environment.lookup_in_env(node.target.name)
+            self.type_constraints.unify(target_tvar, rtype.__args__[0])
+        node.type_constraints = TypeInfo(NoType)
+
+    def visit_listcomp(self, node):
+        # based on the types of the generators.. can set up the type constraints.
+        # the type of the list comprehension should be a list of the elt's types
+        # if the elt is a name node, look up the name in the listcomp node's env and set the type const as that
+        # if it is an operation node,
+        node.type_constraints = TypeInfo(List[self.type_constraints.lookup_concrete(node.type_environment.
+                                                                                        lookup_in_env(node.elt.name))])
```

Comment:
You still aren't thinking recursively here. You should be accessing `node.elt.type_constraints.type`.